src/components/include/utils/message_queue.h

```diff
@@ -171,13 +171,11 @@ bool MessageQueue<T, Q>::IsShuttingDown() const {
 
 template <typename T, class Q>
 void MessageQueue<T, Q>::push(const T& element) {
-  {
-    sync_primitives::AutoLock auto_lock(queue_lock_);
-    if (shutting_down_) {
-      return;
-    }
-    queue_.push(element);
+  sync_primitives::AutoLock auto_lock(queue_lock_);
+  if (shutting_down_) {
+    return;
   }
+  queue_.push(element);
   queue_new_items_.Broadcast();
```

Comment:
It seems to me like this change would cause a DCHECK, I'm surprised that this is supposed to fix an issue. If one thread has called `Wait`, that would mean that when the `Broadcast` is received we would reach `AssertFreeAndMarkTaken`, which will cause a `NOTREACHED` call if the lock is still held.

Reply:
Fix moved from SYNC4 codebase. Additional testing and description will be provided

Reply:
`Broadcast()` will unlock the `queue_lock_` and let the process that is waiting on `Wait()` to lock the `queue_lock_`. 

In another case between `Broadcast()` from thread1 and `Wait()` from thread2 some other thread may acquire `queue_lock_` that will prevent ability for thread2 to squre

Reply:
Confirmed logic from online resources: https://docs.oracle.com/cd/E19455-01/806-5257/sync-27/index.html

> Call pthread_cond_broadcast() under the protection of the same mutex that is used with the condition variable being signaled. Otherwise, the condition variable could be signaled between the test of the associated condition and blocking in pthread_cond_wait(), which can cause an infinite wait.